### PR TITLE
Have the resolver (`Lock`) controller watch package revisions

### LINKED ISSUE #6004: `Lock` controller doesn't own `FunctionRevision`
### What happened?

The `Lock` resolver controller watches provider and configuration revisions, but not function revisions.

The resolver controller collaborates with the package revision controller to make sure all of a package's dependencies are installed:

1. The revision controller adds a new revision to the lock
2. The revision controller returns an error if one or more dependencies don't appear in the lock 
1. The resolver creates installs any missing dependencies 

Watching revisions ensures the resolver controller can quickly install dependencies when a revision is created, allowing the revision controller to requeue and eventually find that its dependencies are installed. So I think function revisions being unwatched means it'd take a long time to install function dependencies - e.g. if a function declared a dependency on a provider.

As a side note, I think we should be using `Watches`, not `Owns`. `Owns` is for types a controller creates and manages the lifecycle of. The resolver controller doesn't create revisions - it creates packages. It does need to watch revisions though (and I don't think it's important that it watches packages).

### How can we reproduce it?

 https://github.com/crossplane/crossplane/blob/878f8a42a4c30a4219fc341f767d003434a62f82/internal/controller/pkg/resolver/reconciler.go#L144-L145

### What environment did it happen in?
Crossplane version: 878f8a42a4c30a4219fc341f767d003434a62f82

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes

Fixes #6004 

We want to enqueue a reconcile of the Lock controller when any package revision changes, but we weren't.

It wasn't watching function revisions at all. It was watching provider and configuration revisions, but using the `Owns()` builder method. This method enqeueus requests for the controller reference of the watched object. The lock isn't the controller reference (or even an owner reference) of any package revision, so we wouldn't be enqueuing reconciles for it.

This PR also includes a commit that fixes saving package metadata CRDs. Nothing really consumes these, so we didn't notice they weren't being saved.

I have: 

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `earthly +reviewable` to ensure this PR is ready for review.
- [ ] ~Added or updated unit tests.~
- [ ] ~Added or updated e2e tests.~
- [ ] ~Linked a PR or a [docs tracking issue] to [document this change].~
- [x] Added `backport release-x.y` labels to auto-backport this PR.

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/main/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/main/contributing#checklist-cheat-sheet

### CODE REVIEW

**Review by jbw976 — APPROVED** [review]
The implementation and reasoning looks reasonable enough. Thanks!!

A couple questions for more context if you have the time:

* Was this a regression from recent package manager changes, or had this been happening for a long time?
* How was the behavior of not enqueuing reconciles for the Lock controller manifesting as user noticeable behavior? would packages take a long time to be updated/upgraded (until a poll caught them)?

**Review by turkenh — APPROVED** [review]

File: internal/controller/pkg/resolver/watch.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+			return true
+		}
+		// Filter ImageConfigs that don't have a pull secret.
+		return ic.Spec.Registry == nil || ic.Spec.Registry.Authentication == nil || ic.Spec.Registry.Authentication.PullSecretRef.Name == ""
+	}
+}
+
+// ForName enqueues a request for the named object. It only enqueues a request
+// if all supplied filter functions return false.
+func ForName(name string, fns ...FilterFn) handler.MapFunc {
```
Comment (turkenh) [inline internal/controller/pkg/resolver/watch.go:48]: Thanks for polish up and unit tests 🙇

**Comment by turkenh** [comment]
This fixes https://github.com/crossplane/crossplane/issues/6004 right?

**Comment by jbw976** [comment]
I just pushed a [commit](https://github.com/crossplane/crossplane/pull/6054/commits/c7b6640c5dd8bf537cb389da1a6eab0f6875c55e) to your branch @negz to hopefully fix the [`check-diff`](https://github.com/crossplane/crossplane/actions/runs/11638148526/job/32412495482?pr=6054) failure, just by running `earthly +generate`. hope that's helpful! 🙇‍♂️

**Comment by negz** [comment]
> A couple questions for more context

I don't really know the answers to either. My guess is it's always been like this. I doubt it's a regression. Impact wise, my best guess is the lack of watch didn't really matter for whatever reason. Perhaps due to polling.

**Pull request merged**